### PR TITLE
Adx 477 access request email org admins

### PR DESCRIPTION
**Pull request by Manoj-nathwani**
# Problem
- Sometimes a dataset does not have a maintainer set, so emails to request access to it kinda go nowhere. Only our internal unattended fjelltopp email is sent an email informing us

# Solution
- Email all org admins which the dataset belongs to notifying them

### CODE REVIEW

File: ckanext/restricted/controller.py [hunk 1]
```diff
@@ -85,6 +85,35 @@ def _send_request_mail(self, data):
...
+                if user['capacity'] == 'admin'
+            ]
+            # fetch users directly from db to get non-hashed emails
+            dataset_org_admins = model.Session\
+                .query(model.User)\
+                .filter(
+                    model.User.id.in_(dataset_org_admins),
+                    model.User.email.isnot(None)
+                )\
+                .all()
```
Comment (tomeksabala) [inline ckanext/restricted/controller.py:111]: Usually we use line breaks without \
sth like:
```
model.Session.query(model.User).filter(
                    model.User.id.in_(dataset_org_admins),
                    model.User.email.isnot(None)
).all()
```

File: ckanext/restricted/controller.py [hunk 2]
```diff
@@ -85,6 +85,35 @@ def _send_request_mail(self, data):
...
+                'session': model.Session,
+                'ignore_auth': True
+            }
+            dataset = toolkit.get_action('package_show')(
+                context, {'id': data.get('package_name')}
+            )
+            dataset_org = toolkit.get_action('organization_show')(
+                context, {'id': dataset['owner_org']}
+            )
+            dataset_org_admins = [
```
Comment (tomeksabala) [inline ckanext/restricted/controller.py:99]: please change the var name to `dataset_org_admin_ids`.
It's generally a bad practice to mutate variables if you don't have to (this var is reassigned a new value in :105)

File: ckanext/restricted/tests/test_plugin.py [hunk 3]
```diff
@@ -436,3 +438,64 @@ def test_restricted_string_snippet(self):
...
+
+        mocked_mail_recipient.assert_called()
+        email_recipients = [
+            x[0][1]  # recipient_email
+            for x in mocked_mail_recipient.call_args_list
+        ]
+        assert maintainer_email in email_recipients
+        assert user_1['email'] in email_recipients
+        assert user_2['email'] in email_recipients
+        assert user_3['email'] not in email_recipients, \
```
Comment (tomeksabala) [inline ckanext/restricted/tests/test_plugin.py:500]: why a line break here? We have set a line length to 127

File: ckanext/restricted/tests/test_plugin.py [hunk 4]
```diff
@@ -436,3 +438,64 @@ def test_restricted_string_snippet(self):
...
         )
         assert expected in html4
+
+    @mock.patch('ckan.lib.mailer.mail_recipient')
+    def test_all_org_admins_are_emailed_on_request_access(self, mocked_mail_recipient, app):
+
+        # create two admins and one regular member of an org
+        user_1 = factories.User(email='user_1@example.com')
+        user_2 = factories.User(email='user_2@example.com')
+        user_3 = factories.User(email='user_3@example.com')
```
Comment (tomeksabala) [inline ckanext/restricted/tests/test_plugin.py:448]: if you name the variables differently the comment "# create two admins and one regular member of an org" won't be needed
```
admin1 = factories
admin2 = factories
member = factories
```

**Review by tomeksabala — CHANGES_REQUESTED** [review]
A few small changes.

I also think it'd be great to add a few more tests for this:
`test_both_org_admin_and_maintainer_are_emailed_on_request_access` - single admin in org and maintainer provided
`test_org_admin_is_emailed_if_no_maintaner_on_request_access` - a single admin and no maintainer info provided
`test_only_org_admins_are_emailed_on_request_access` - an admin, editor, and member in the owner org and only the admin is e-mailed.

**Comment by Manoj-nathwani** [comment]
I've amended the tests, there's no need to create more as it would mean we'd duplicate a lot of code.

I'm testing:
- dataset maintainer is emailed
- dataset org admins get emails
- dataset org editors & members do not

**Review by tomeksabala — CHANGES_REQUESTED** [review]
Great, the code looks good now. About the tests however: I'd suggest going with my initial suggestion and creating 4 separate test methods. Don't worry about duplication (if you see any, just extract common logic into a function. This can be done for test setup (given), action (when) and even assert (then)). Having those 4 separate test cases is going to do better job at testing the feature which a single "do it all" test does not. Also it improves the readability a lot in case of a failing test report.
I'm happy to explain it more in a call if it's not clear.

**Comment by Manoj-nathwani** [comment]
@tomeksabala apologies for the delay - this is now ready to be merged in 🚀

**Review by tomeksabala — APPROVED** [review]
"This is the way", LGTM :+1:

**Pull request merged**